### ISSUE TITLE
How To Do Activity Maximization In Caffe

### ISSUE THREAD

**Issue by minhtannguyen** [BVLC/caffe#833]
Hi everyone,

I am trying to do activity maximization described in Karen Simonyan's paper (http://arxiv.org/pdf/1312.6034.pdf) using the pre-trained ImageNet model in Caffe. I want to find the input image that maximize the score of each class in the Softmax layer. In other to do so, derivatives of the scores versus input image need to be computed. Do you know how to use back propagation to calculate those derivatives in Caffe?

Thank you so much!!!

Sincerely,

Tan Nguyen

**Comment by shelhamer** [comment]
Please ask on the [caffe-users](https://groups.google.com/forum/#!forum/caffe-users) mailing list. Thanks!

**Comment by dk683** [comment]
Hello. Could I ask for update @minhtannguyen and @resterhall?
I have the following files: 
deploy.prototxt
```
name: "CaffeNet"

input: "data"
input_dim: 1
input_dim: 3
input_dim: 227
input_dim: 227

force_backward: true

layers {
  name: "conv1"
  type: CONVOLUTION
  bottom: "data"
  top: "conv1"
  convolution_param {
    num_output: 96
    kernel_size: 11
    stride: 4
  }
}

layers {
  name: "relu1"
  type: RELU
  bottom: "conv1"
  top: "conv1"
}

layers {
  name: "pool1"
  type: POOLING
  bottom: "conv1"
  top: "pool1"
  pooling_param {
    pool: MAX
    kernel_size: 3
    stride: 2
  }
}

layers {
  name: "norm1"
  type: LRN
  bottom: "pool1"
  top: "norm1"
  lrn_param {
    local_size: 5
    alpha: 0.0001
    beta: 0.75
  }
}

layers {
  name: "conv2"
  type: CONVOLUTION
  bottom: "norm1"
  top: "conv2"
  convolution_param {
    num_output: 256
    pad: 2
    kernel_size: 5
    group: 2
  }
}
```

deconvnet.py
```python
# IMPORT LIBRARY
import numpy as np
import matplotlib.pyplot as plt
import sys, caffe, os, operator
import caffe.io
from caffe.proto import caffe_pb2
from PIL import Image

# FORWARDVISUALIZE FUNCTION
# Visualize class vs probability
def classVisualize (forwardOutput, blobName):
	reshapedArray = np.reshape(forwardOutput[blobName][0], (len(forwardOutput[blobName][0]),1))
	plt.plot(reshapedArray)
	plt.show()

# VISSQUARE
# take an array of shape (n, height, width) or (n, height, width, channels)
# and visualize each (height, width) thing in a grid of size approx. sqrt(n) by sqrt(n)
def visSquare(data, padsize=1, padval=0):
    data -= data.min()
    data /= data.max()
    
    # force the number of filters to be square
    n = int(np.ceil(np.sqrt(data.shape[0])))
    padding = ((0, n ** 2 - data.shape[0]), (0, padsize), (0, padsize)) + ((0, 0),) * (data.ndim - 3)
    data = np.pad(data, padding, mode='constant', constant_values=(padval, padval))
    
    # tile the filters into an image
    data = data.reshape((n, n) + data.shape[1:]).transpose((0, 2, 1, 3) + tuple(range(4, data.ndim + 1)))
    data = data.reshape((n * data.shape[1], n * data.shape[3]) + data.shape[4:])
    
    plt.imshow(data)
    plt.show()

# MAIN FUNCTION
# Define variables and path
modelFile = '/home/caffe-master/examples/deconvnet/deploy.prototxt'
trainedModel = '/home/caffe-master/examples/imagenet/caffe_reference_imagenet_model'
meanFile = '/home/caffe-master/python/caffe/imagenet/ilsvrc_2012_mean.npy'
imageFile = '/home/caffe-master/examples/images/cat.jpg'

# Create net
caffe.set_mode_gpu()
net = caffe.Classifier(modelFile, trainedModel)
net.set_mean('data', np.load(meanFile))
net.set_raw_scale('data', 255)
net.set_channel_swap('data', (2,1,0))

# Display Net structure
print 'Net structure:'
for k,v in net.blobs.items():
	print (k, v.data.shape)

# Define input image to net
inputImage = caffe.io.load_image(imageFile)
inputResized = np.reshape(caffe.io.resize_image(inputImage, net.image_dims),[1,227,227,3])
caffeInput = np.asarray([net.preprocess('data', in_) for in_ in inputResized])

# Forward
forwardOutput = net.forward(**{net.inputs[0]: caffeInput})

# Backward: compute diff
# backwardOutput['data'].shape: (1,3,227,227)
backwardOutput = net.backward(**{net.outputs[0]: forwardOutput['conv2']})
visSquare(backwardOutput['data'].transpose(0, 2, 3, 1))
```
However, the result does not have a good quality as the result in the paper (http://arxiv.org/pdf/1312.6034.pdf). Could you help me?

Also, the paper says "We can conclude that apart from the RELU layer, computing the approximate feature map reconstruction R_n using a DeconvNet is equivalent to computing the derivative ∂f /∂Xn using backpropagation, which is a part of our visualisation algorithms."
However, I think the code is doing backpropagation not only conv and pooling layer, but also relu layer. Does this affect result? 
I appreciate your time!

**Comment by smajida** [comment]
Hi there,

Thanks @minhtannguyen for your expalanation, I've been trying to visualize the figure 2 for one whole week could you please tell me how did you manage to do that? I've got some results but not sure which is true?
